cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LocalizationUtils.java

```diff
@@ -0,0 +1,152 @@
+ */
+public final class LocalizationUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalizationUtils.class);
+
+  /**
+   * Localizes the specified {@link LocalizeResource} in the specified {@link File targetDir} with the specified
+   * file name and returns the {@link File} pointing to the localized file.
+   *
+   * @param fileName the name to localize the file with
+   * @param resource the {@link LocalizeResource} to localize
+   * @param targetDir the directory to localize the resource in
+   * @return the {@link File} pointing to the localized file.
+   */
+  public static File localizeResource(String fileName, LocalizeResource resource, File targetDir) throws IOException {
+    File localizedResource = new File(targetDir, fileName);
+    File input = getFileToLocalize(resource);
+    if (resource.isArchive()) {
+      LOG.debug("Decompress file {} to {}", input, localizedResource);
+      unpack(input, localizedResource);
+    } else {
+      try {
+        LOG.debug("Hard link file from {} to {}", input, localizedResource);
+        java.nio.file.Files.createLink(Paths.get(localizedResource.toURI()), Paths.get(input.toURI()));
+      } catch (Exception e) {
+        LOG.debug("Copy file from {} to {}", input, localizedResource);
+        Files.copy(input, localizedResource);
+      }
+    }
+    return localizedResource;
+  }
+
+  /**
+   * Returns a local {@link File} for the specified {@link LocalizeResource}. If the specified {@link LocalizeResource}
+   * points to a local file already, it simply returns a {@link File} object for it. If it is not local, then the
+   * method will try to download the resource to a local temporary file and return it.
+   *
+   * @param resource the {@link LocalizeResource} for which a local file is requested
+   * @return a local {@link File} for the specified resource
+   */
+  private static File getFileToLocalize(LocalizeResource resource) throws IOException {
+    URI uri = resource.getURI();
+    if ("file".equals(uri.getScheme())) {
+      // Local file. Just return a File object for the file.
+      return new File(uri.getPath());
+    }
+    URL url = uri.toURL();
+    String name = new File(uri.getPath()).getName();
+    String extension = Files.getFileExtension(name);
+    if (!extension.isEmpty()) {
+      name = name.substring(0, name.indexOf(extension));
+    }
+    File tempFile = File.createTempFile(name, extension);
+    Files.copy(Resources.newInputStreamSupplier(url), tempFile);
+    return tempFile;
+  }
+
+  private static void unpack(File archive, File targetDir) throws IOException {
+    if (!targetDir.exists()) {
+      //noinspection ResultOfMethodCallIgnored
+      targetDir.mkdir();
+    }
+    String extension = Files.getFileExtension(archive.getPath()).toLowerCase();
+    switch (extension) {
+      case "zip":
+      case "jar":
+        BundleJarUtil.unJar(Locations.toLocation(archive), targetDir);
+        break;
+      case "tgz":
+      case "gz":
```

Comment:
I was counting on the untaring to fail. However, I agree, added a precondition to throw a meaningful error now if a ``.gz`` file is passed.